In [1]:
import pandas as pd
import numpy as np
# import matplotlib.pyplot as plt
from pandas import DataFrame

# Key concept to know 
`pd.read_cs`           - df = pd.read_csv('myfile.csv')
`df.to_csv`            - df.to_csv('myfile.csv')
`pd.read_json`         - df = pd.read_json('myfile.json')
`df.corr`              - df.corr()
`df.dropna`            - df.dropna()
`df.loc`               - df.loc[df['trip_distance'] > 10, 'passenger_count'] 
`pd.read_html`         - df = df.read.html('https://a-site.)
`s.value_counts`       - s.value_counts() 
`s.round`              - s.round(2)     
`df.memory_usage`      - df.memory_usage()
`pd.Series.idxmin`     - s.idxmin()
`pd.Series.idxmax`     - s.idxmax()
`pd.DatFrame.agg`      - df.idmax(['min','max'])

In [2]:
# arguments `pd.read_csv` & `df.to_csv`: 
# `sep`the field separator 
# `header` whether ther are headers sescribing column names and on which line of the file they appear
# `index_col` which column, if any, should be set to be the index of our data frame
# `usecols` which column from the file should be included in the data frame  


# Exercise 15: Weird taxi rides


In [3]:
ex15 = pd.read_csv(r"C:\Users\dqthi\Downloads\pandas-workout-data\data\nyc_taxi_2019-01.csv",
                     usecols = ['passenger_count', 'trip_distance', 'payment_type', 'total_amount'])

In [4]:
ex15.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7667792 entries, 0 to 7667791
Data columns (total 4 columns):
 #   Column           Dtype  
---  ------           -----  
 0   passenger_count  int64  
 1   trip_distance    float64
 2   payment_type     int64  
 3   total_amount     float64
dtypes: float64(2), int64(2)
memory usage: 234.0 MB


In [5]:
# passenger_count—The number of passengers who took that taxi ride.

# trip_distance—The distance traveled in miles.

# total_amount—The total owed to the driver, including the fare, tolls, taxes, and tip.

# payment_type—An integer number describing how the passenger paid for the trip. The most important values are 1 (credit card) and 2 (cash).

In [6]:
ex15.shape 

(7667792, 4)

In [7]:
# there are 7.6 million rows in the dataset. this is massive. i think this is the first time I am processing the data with this amount of rows. ``

In [8]:
ex15.describe()

passenger_count  trip_distance  payment_type  total_amount
count     7.667792e+06   7.667792e+06  7.667792e+06  7.667792e+06
mean      1.567078e+00   2.801084e+00  1.291776e+00  1.568222e+01
std       1.224431e+00   3.737529e+00  4.733229e-01  2.622932e+02
min       0.000000e+00   0.000000e+00  1.000000e+00 -3.628000e+02
25%       1.000000e+00   9.000000e-01  1.000000e+00  8.190000e+00
50%       1.000000e+00   1.530000e+00  1.000000e+00  1.127000e+01
75%       2.000000e+00   2.800000e+00  2.000000e+00  1.656000e+01
max       9.000000e+00   8.318000e+02  4.000000e+00  6.232617e+05

In [9]:
# Q: How many taxi rides had more than eight passengers?
# A: Out 7.67 million rides, only 9 rides are more than 8 passengers 

In [10]:
ex15.loc[ex15['passenger_count'] > 8 ].shape

(9, 4)

In [11]:
# Q: How many taxi rides had zero passengers?
# A: out of 7.67million rides, there is 117k that has no passenger 

In [12]:
ex15.loc[ex15['passenger_count'] == 0].shape

(117381, 4)

In [13]:
# Q: How many taxi rides were paid for in cash and cost over $1,000?
# A: Out of 7.67 miilion rides, there are 5 rides that paid in cash with the amount over $1,000  

In [14]:
ex15.loc[(ex15['payment_type'] == 2) & (ex15['total_amount'] > 1000)].info() 

<class 'pandas.core.frame.DataFrame'>
Index: 5 entries, 478791 to 6827160
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   passenger_count  5 non-null      int64  
 1   trip_distance    5 non-null      float64
 2   payment_type     5 non-null      int64  
 3   total_amount     5 non-null      float64
dtypes: float64(2), int64(2)
memory usage: 200.0 bytes


In [15]:
# Q: How many rides cost less than $0?  
# A: There are 7131 rides that cost less than $0 

In [16]:
ex15.loc[ex15['total_amount'] < 0].shape

(7131, 4)

In [17]:
# Q: How many rides traveled a below-average distance but cost an above-average amount?
# A: 411,255 rides that are below-average distance and cost an above-average amount 

In [18]:
ex15['total_amount'].mean()

15.68222215991253

In [19]:
ex15['trip_distance'].mean()

2.8010838491706607

In [20]:
ex15.loc[(ex15['trip_distance'] < ex15['trip_distance'].mean()) & ( ex15['total_amount'] > ex15['total_amount'].mean())].info()

<class 'pandas.core.frame.DataFrame'>
Index: 411255 entries, 1 to 7667754
Data columns (total 4 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   passenger_count  411255 non-null  int64  
 1   trip_distance    411255 non-null  float64
 2   payment_type     411255 non-null  int64  
 3   total_amount     411255 non-null  float64
dtypes: float64(2), int64(2)
memory usage: 15.7 MB


In [21]:
# bte: 
# Repeat this exercise using the query method rather than a boolean index and loc.


In [22]:
ex15.query('passenger_count > 8').count()

passenger_count    9
trip_distance      9
payment_type       9
total_amount       9
dtype: int64

In [23]:
ex15.query('passenger_count == 0').count()

passenger_count    117381
trip_distance      117381
payment_type       117381
total_amount       117381
dtype: int64

In [24]:
ex15.query('(payment_type == 2) & (total_amount > 1000)').count()

passenger_count    5
trip_distance      5
payment_type       5
total_amount       5
dtype: int64

In [25]:
ex15.query('total_amount < 0').count()

passenger_count    7131
trip_distance      7131
payment_type       7131
total_amount       7131
dtype: int64

In [26]:
ex15.query('(trip_distance < trip_distance.mean()) & (total_amount > total_amount.mean())').count()

passenger_count    411255
trip_distance      411255
payment_type       411255
total_amount       411255
dtype: int64

In [27]:
# bte: 
# How many rides that cost less than $0 involved either 
# a dispute (payment_type of 4) or a voided trip (payment_type of 6)?
# Answer: 2666 rides that had payment type are either dispute or voided with the total costs less than 0$0

In [28]:
ex15.loc[(ex15['total_amount'] < 0) & ((ex15['payment_type'] == 4) | (ex15['payment_type'] == 6))].shape

(2666, 4)

In [29]:
# bte: 
# Q: show rides with payments as credit cards and then calculate the percentage of payments by credit cards over total rides

In [30]:
ex15.loc[ex15['payment_type'] == 1]['payment_type'].sum() / ex15['payment_type'].count() * 100

71.54637214989661

In [31]:
payment_type_table = ex15.pivot_table(ex15, columns=['payment_type'],aggfunc='count')
payment_type_table

payment_type           1        2      3      4
passenger_count  5486027  2137415  33186  11164
total_amount     5486027  2137415  33186  11164
trip_distance    5486027  2137415  33186  11164

In [32]:
payment_type_table2 = ex15.pivot_table(ex15, index=['payment_type'], aggfunc='count')
payment_type_table2

passenger_count  total_amount  trip_distance
payment_type                                              
1                     5486027       5486027        5486027
2                     2137415       2137415        2137415
3                       33186         33186          33186
4                       11164         11164          11164

# Exercise 16: Pandemic taxis 

In [33]:
# Questions:
# How many rides were taken in 2019 and 2020, and what is the difference between these two figures?

# How much money (in total) was collected in 2019 and 2020, and what was the difference between these two figures?

# Did the proportion of trips with more than one passenger change dramatically?

# Did people use cash (i.e., payment_type of 2) less in 2020 than in 2019?

In [34]:
df_2019_jul = pd.read_csv(r"C:\Users\dqthi\Downloads\pandas-workout-data\data\nyc_taxi_2019-07.csv",
                            usecols=['passenger_count', 'total_amount', 'payment_type'])

In [35]:
df_2019_jul.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6310419 entries, 0 to 6310418
Data columns (total 3 columns):
 #   Column           Dtype  
---  ------           -----  
 0   passenger_count  float64
 1   payment_type     float64
 2   total_amount     float64
dtypes: float64(3)
memory usage: 144.4 MB


In [36]:
df_2020_jul = pd.read_csv(r"C:\Users\dqthi\Downloads\pandas-workout-data\data\nyc_taxi_2020-07.csv",
                          usecols = ['passenger_count', 'total_amount', 'payment_type'])

In [37]:
df_2020_jul.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800412 entries, 0 to 800411
Data columns (total 3 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   passenger_count  737565 non-null  float64
 1   payment_type     737565 non-null  float64
 2   total_amount     800412 non-null  float64
dtypes: float64(3)
memory usage: 18.3 MB


In [38]:
ex16 = pd.concat([df_2019_jul, df_2020_jul])

In [39]:
ex16

passenger_count  payment_type  total_amount
0                   1.0           1.0          4.94
1                   1.0           2.0         20.30
2                   1.0           1.0         70.67
3                   1.0           1.0         66.36
4                   0.0           1.0         15.30
...                 ...           ...           ...
800407              NaN           NaN         83.50
800408              NaN           NaN         19.78
800409              NaN           NaN         38.45
800410              NaN           NaN         29.77
800411              NaN           NaN         51.90

[7110831 rows x 3 columns]

In [40]:
ex16.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7110831 entries, 0 to 800411
Data columns (total 3 columns):
 #   Column           Dtype  
---  ------           -----  
 0   passenger_count  float64
 1   payment_type     float64
 2   total_amount     float64
dtypes: float64(3)
memory usage: 217.0 MB


In [41]:
df_2019_jul['year'] = 2019

In [42]:
df_2020_jul['year'] = 2020

In [43]:
df_2019_jul.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6310419 entries, 0 to 6310418
Data columns (total 4 columns):
 #   Column           Dtype  
---  ------           -----  
 0   passenger_count  float64
 1   payment_type     float64
 2   total_amount     float64
 3   year             int64  
dtypes: float64(3), int64(1)
memory usage: 192.6 MB


In [44]:
df_2020_jul.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800412 entries, 0 to 800411
Data columns (total 4 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   passenger_count  737565 non-null  float64
 1   payment_type     737565 non-null  float64
 2   total_amount     800412 non-null  float64
 3   year             800412 non-null  int64  
dtypes: float64(3), int64(1)
memory usage: 24.4 MB


In [45]:
ex16 = pd.concat([df_2019_jul, df_2020_jul])
ex16

passenger_count  payment_type  total_amount  year
0                   1.0           1.0          4.94  2019
1                   1.0           2.0         20.30  2019
2                   1.0           1.0         70.67  2019
3                   1.0           1.0         66.36  2019
4                   0.0           1.0         15.30  2019
...                 ...           ...           ...   ...
800407              NaN           NaN         83.50  2020
800408              NaN           NaN         19.78  2020
800409              NaN           NaN         38.45  2020
800410              NaN           NaN         29.77  2020
800411              NaN           NaN         51.90  2020

[7110831 rows x 4 columns]

In [46]:
ex16.loc[ex16['year'] == 2020] 

passenger_count  payment_type  total_amount  year
0                   1.0           2.0          9.30  2020
1                   1.0           1.0         27.80  2020
2                   1.0           2.0         22.30  2020
3                   1.0           1.0         14.16  2020
4                   1.0           2.0          7.80  2020
...                 ...           ...           ...   ...
800407              NaN           NaN         83.50  2020
800408              NaN           NaN         19.78  2020
800409              NaN           NaN         38.45  2020
800410              NaN           NaN         29.77  2020
800411              NaN           NaN         51.90  2020

[800412 rows x 4 columns]

In [47]:
ex16.loc[ex16['year'] == 2019] 

passenger_count  payment_type  total_amount  year
0                    1.0           1.0          4.94  2019
1                    1.0           2.0         20.30  2019
2                    1.0           1.0         70.67  2019
3                    1.0           1.0         66.36  2019
4                    0.0           1.0         15.30  2019
...                  ...           ...           ...   ...
6310414              NaN           NaN         29.00  2019
6310415              NaN           NaN         54.43  2019
6310416              NaN           NaN         65.40  2019
6310417              NaN           NaN         43.00  2019
6310418              NaN           NaN         22.15  2019

[6310419 rows x 4 columns]

In [48]:
(
    ex16.loc[ex16['year'] == 2019, 'total_amount'].count() - 
    ex16.loc[ex16['year'] == 2020, 'total_amount'].count()
)

5510007

In [49]:
diff_2019_2020 = (
    ex16.loc[ex16['year'] == 2019, 'total_amount'].count() - 
    ex16.loc[ex16['year'] == 2020, 'total_amount'].count()
)

In [50]:
diff_2019_2020

5510007

In [51]:
ex16.loc[ex16['year'] == 2019, ['total_amount', 'passenger_count']]

total_amount  passenger_count
0                4.94              1.0
1               20.30              1.0
2               70.67              1.0
3               66.36              1.0
4               15.30              0.0
...               ...              ...
6310414         29.00              NaN
6310415         54.43              NaN
6310416         65.40              NaN
6310417         43.00              NaN
6310418         22.15              NaN

[6310419 rows x 2 columns]

In [52]:
ex16.loc[ex16['year'] == 2019][['total_amount','passenger_count']]

total_amount  passenger_count
0                4.94              1.0
1               20.30              1.0
2               70.67              1.0
3               66.36              1.0
4               15.30              0.0
...               ...              ...
6310414         29.00              NaN
6310415         54.43              NaN
6310416         65.40              NaN
6310417         43.00              NaN
6310418         22.15              NaN

[6310419 rows x 2 columns]

In [53]:
ex16[ex16['year'] == 2019][['total_amount','passenger_count']]

total_amount  passenger_count
0                4.94              1.0
1               20.30              1.0
2               70.67              1.0
3               66.36              1.0
4               15.30              0.0
...               ...              ...
6310414         29.00              NaN
6310415         54.43              NaN
6310416         65.40              NaN
6310417         43.00              NaN
6310418         22.15              NaN

[6310419 rows x 2 columns]

In [54]:
ex16.loc[(ex16['year'] == 2019) & (ex16['passenger_count'] != 1)]['year'].count()

1929529

In [55]:
ex16.loc[(ex16['year'] == 2020) & (ex16['passenger_count'] != 1)]['year'].count()

234403

In [56]:
pass_more_than_1_diff = (ex16.loc[(ex16['year'] == 2019) & (ex16['passenger_count'] != 1)]['year'].count() - 
                        ex16.loc[(ex16['year'] == 2020) & (ex16['passenger_count'] != 1)]['year'].count())

In [57]:
pass_more_than_1_diff

1695126

In [58]:
(ex16.loc[(ex16['year'] == 2019) & (ex16['passenger_count'] > 1)]['passenger_count'].count()) / (ex16.loc[(ex16['year'] == 2019)]['passenger_count'].count()) * 100

28.339000009559527

In [59]:
(ex16.loc[(ex16['year'] == 2020) & (ex16['passenger_count'] > 1)]['passenger_count'].count()) / (ex16.loc[(ex16['year'] == 2020)]['passenger_count'].count()) * 100

20.61513222563435

In [60]:
(ex16.loc[(ex16['year'] == 2019) & (ex16['payment_type'] == 2)]['payment_type'].count()) / (ex16.loc[(ex16['year'] == 2019)]['payment_type'].count()) * 100

28.705958454287927

In [61]:
(ex16.loc[(ex16['year'] == 2020) & (ex16['payment_type'] == 2)]['payment_type'].count()) / (ex16.loc[(ex16['year'] == 2020)]['payment_type'].count()) * 100

32.0558865998251

In [62]:
ex16.corr()

passenger_count  payment_type  total_amount      year
passenger_count         1.000000      0.016410      0.014943 -0.049558
payment_type            0.016410      1.000000     -0.138561  0.029277
total_amount            0.014943     -0.138561      1.000000 -0.019706
year                   -0.049558      0.029277     -0.019706  1.000000

In [63]:
ex16.loc[ex16['year'] == 2019, 'total_amount'].describe().round(2)

count    6310419.00
mean          19.61
std           15.81
min         -450.80
25%           11.30
50%           14.76
75%           21.30
max         6667.45
Name: total_amount, dtype: float64

In [64]:
ex16.loc[ex16['year'] == 2020, 'total_amount'].describe().round(2)

count    800412.00
mean         18.63
std          15.06
min        -397.60
25%          10.80
50%          14.16
75%          20.55
max        1995.00
Name: total_amount, dtype: float64

In [65]:
ex16.loc[ex16['passenger_count'] == 2019,   'total_amount'].describe().round(2)

count    0.0
mean     NaN
std      NaN
min      NaN
25%      NaN
50%      NaN
75%      NaN
max      NaN
Name: total_amount, dtype: float64

# Exercise 17: Setting column types 

### Specify the dtype for each column as you read it in.

In [66]:
filepath = r"C:\Users\dqthi\Downloads\pandas-workout-data\data\nyc_taxi_2020-01.csv"

ex17_df = pd.read_csv(filepath)

C:\Users\dqthi\AppData\Local\Temp\ipykernel_50028\1989683883.py:3: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  ex17_df = pd.read_csv(filepath)


In [67]:
ex17_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6405008 entries, 0 to 6405007
Data columns (total 18 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   VendorID               float64
 1   tpep_pickup_datetime   object 
 2   tpep_dropoff_datetime  object 
 3   passenger_count        float64
 4   trip_distance          float64
 5   RatecodeID             float64
 6   store_and_fwd_flag     object 
 7   PULocationID           int64  
 8   DOLocationID           int64  
 9   payment_type           float64
 10  fare_amount            float64
 11  extra                  float64
 12  mta_tax                float64
 13  tip_amount             float64
 14  tolls_amount           float64
 15  improvement_surcharge  float64
 16  total_amount           float64
 17  congestion_surcharge   float64
dtypes: float64(13), int64(2), object(3)
memory usage: 879.6+ MB


#### Observation: 
- memory usage takes upto 879.6 mb to load the file  
- receive a error message "C:\Users\dqthi\AppData\Local\Temp\ipykernel_60784\1989683883.py:3: DtypeWarning: Columns (6) have mixed types. Specify dtype option on import or set low_memory=False.
  ex17_df = pd.read_csv(filepath)"  
      - *how do I know if which columns are having mixed types*

In [68]:
filepath = r"C:\Users\dqthi\Downloads\pandas-workout-data\data\nyc_taxi_2020-01.csv"

ex17b_df = pd.read_csv(filepath, 
                       usecols =['passenger_count', 'payment_type', 'total_amount'],
                       dtype={'passenger_count':np.float32, 'total_amount':np.float32, 'payment_type':np.float32}
                        )

In [69]:
ex17b_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6405008 entries, 0 to 6405007
Data columns (total 3 columns):
 #   Column           Dtype  
---  ------           -----  
 0   passenger_count  float32
 1   payment_type     float32
 2   total_amount     float32
dtypes: float32(3)
memory usage: 73.3 MB



### Identify rows containing NaN values. Which columns are NaN, and why?





Remove any rows containing any NaN values.

Set the dtype for each column to the smallest, most appropriate value.

In [70]:
ex17b_df.isnull().sum()

passenger_count    65441
payment_type       65441
total_amount           0
dtype: int64

In [71]:
ex17b_df_clean = ex17b_df.dropna().copy()

In [72]:
ex17b_df_clean

passenger_count  payment_type  total_amount
0                    1.0           1.0     11.270000
1                    1.0           1.0     12.300000
2                    1.0           1.0     10.800000
3                    1.0           1.0      8.160000
4                    1.0           2.0      4.800000
...                  ...           ...           ...
6339562              1.0           1.0     17.760000
6339563              1.0           1.0     20.160000
6339564              1.0           1.0     19.559999
6339565              1.0           2.0     12.300000
6339566              1.0           1.0      0.000000

[6339567 rows x 3 columns]

In [73]:
ex17b_df_clean['passenger_count'] = ex17b_df_clean['passenger_count'].astype(np.int8)
ex17b_df_clean

passenger_count  payment_type  total_amount
0                      1           1.0     11.270000
1                      1           1.0     12.300000
2                      1           1.0     10.800000
3                      1           1.0      8.160000
4                      1           2.0      4.800000
...                  ...           ...           ...
6339562                1           1.0     17.760000
6339563                1           1.0     20.160000
6339564                1           1.0     19.559999
6339565                1           2.0     12.300000
6339566                1           1.0      0.000000

[6339567 rows x 3 columns]

In [ ]:
ex17b_df_clean['passenger_count'].info() 

<class 'pandas.core.series.Series'>
Index: 6339567 entries, 0 to 6339566
Series name: passenger_count
Non-Null Count    Dtype
--------------    -----
6339567 non-null  int8 
dtypes: int8(1)
memory usage: 54.4 MB


In [ ]:
ex17b_df_clean['payment_type'].info()

<class 'pandas.core.series.Series'>
Index: 6339567 entries, 0 to 6339566
Series name: payment_type
Non-Null Count    Dtype  
--------------    -----  
6339567 non-null  float32
dtypes: float32(1)
memory usage: 72.6 MB


In [ ]:
ex17b_df_clean['payment_type'] = ex17b_df_clean['payment_type'].astype(np.int8)
ex17b_df_clean['payment_type'].info()

<class 'pandas.core.series.Series'>
Index: 6339567 entries, 0 to 6339566
Series name: payment_type
Non-Null Count    Dtype
--------------    -----
6339567 non-null  int8 
dtypes: int8(1)
memory usage: 54.4 MB


### btx: 
Create a data frame from four other columns (VendorID, trip_distance, tip_amount, and total_amount), specifying the dtype for each. What types are most appropriate? Can you use them directly, or must you first clean the data?

In [77]:
ex17_1_df = pd.read_csv(filepath, 
                        usecols=['VendorID', 'trip_distance', 'tip_amount', 'total_amount', 'payment_type'],
                        # dtype={'VendorID':np.int8, 'trip_distance':np.float32, 'tip_amount':np.float32, 'total_amount':np.float32},
#                        na_values=['NA', '']
                        )
ex17_1_df 

VendorID  trip_distance  payment_type  tip_amount  total_amount
0             1.0           1.20           1.0        1.47         11.27
1             1.0           1.20           1.0        1.50         12.30
2             1.0           0.60           1.0        1.00         10.80
3             1.0           0.80           1.0        1.36          8.16
4             2.0           0.00           2.0        0.00          4.80
...           ...            ...           ...         ...           ...
6405003       NaN           3.24           NaN        0.00         21.14
6405004       NaN          22.13           NaN        0.00         62.46
6405005       NaN          10.51           NaN        0.00         51.90
6405006       NaN           5.49           NaN        0.00         30.22
6405007       NaN          11.60           NaN        0.00         58.11

[6405008 rows x 5 columns]

In [78]:
# note: 
# apparently I cannt set the dtypes of a column if there is missing value, so what is the solution? 
# solution 1: 
    # load the data itself then remove the 'NaN' value, then set the type 

# solution 2: 
    # Dont know yet 

In [79]:
ex17_1_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6405008 entries, 0 to 6405007
Data columns (total 5 columns):
 #   Column         Dtype  
---  ------         -----  
 0   VendorID       float64
 1   trip_distance  float64
 2   payment_type   float64
 3   tip_amount     float64
 4   total_amount   float64
dtypes: float64(5)
memory usage: 244.3 MB


In [80]:
ex17_1_df.isnull().sum() 

VendorID         65441
trip_distance        0
payment_type     65441
tip_amount           0
total_amount         0
dtype: int64

In [81]:
ex17_1b_df = ex17_1_df.dropna()

In [82]:
ex17_1b_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6339567 entries, 0 to 6339566
Data columns (total 5 columns):
 #   Column         Dtype  
---  ------         -----  
 0   VendorID       float64
 1   trip_distance  float64
 2   payment_type   float64
 3   tip_amount     float64
 4   total_amount   float64
dtypes: float64(5)
memory usage: 290.2 MB


In [83]:
ex17_1b_df['VendorID'] = ex17_1b_df['VendorID'].astype(np.int8) 
ex17_1b_df['payment_type'] = ex17_1b_df['payment_type'].astype(np.int8) 

C:\Users\dqthi\AppData\Local\Temp\ipykernel_50028\1588491516.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ex17_1b_df['VendorID'] = ex17_1b_df['VendorID'].astype(np.int8)
C:\Users\dqthi\AppData\Local\Temp\ipykernel_50028\1588491516.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ex17_1b_df['payment_type'] = ex17_1b_df['payment_type'].astype(np.int8)


In [84]:
ex17_1b_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6339567 entries, 0 to 6339566
Data columns (total 5 columns):
 #   Column         Dtype  
---  ------         -----  
 0   VendorID       int8   
 1   trip_distance  float64
 2   payment_type   int8   
 3   tip_amount     float64
 4   total_amount   float64
dtypes: float64(3), int8(2)
memory usage: 205.6 MB


In [85]:
# by changing the datatype of 2 columns `VendorID` and `payment_type`, the memory usage consumstion went downt to 205.6MB from 290.2MB 

# Exercise 18: passwd to df 

In [86]:
filepath_ex18 = r"C:\Users\dqthi\Downloads\pandas-workout-data\data\linux-etc-passwd.txt"

ex18_df = pd.read_csv(filepath_ex18,
                sep = ':',
                comment= '#',
                header = None, 
                names=['username', 'password', 'userid', 'groupid', 'name',
                'homedir', 'shell'])

In [87]:
ex18_df

username password  userid  groupid  \
0                 root        x       0        0   
1               daemon        x       1        1   
2                  bin        x       2        2   
3                  sys        x       3        3   
4                 sync        x       4    65534   
5                games        x       5       60   
6                  man        x       6       12   
7                   lp        x       7        7   
8                 mail        x       8        8   
9                 news        x       9        9   
10                uucp        x      10       10   
11               proxy        x      13       13   
12            www-data        x      33       33   
13              backup        x      34       34   
14                list        x      38       38   
15                 irc        x      39       39   
16               gnats        x      41       41   
17              nobody        x   65534    65534   
18              syslog        x     101      104   
19          messagebus        x     102      106   
20           landscape        x     103      109   
21                 jci        x     955      955   
22                sshd        x     104    65534   
23                user        x    1000     1000   
24              reuven        x    1001     1001   
25             postfix        x     105      113   
26              colord        x     106      116   
27            postgres        x     107      117   
28             dovecot        x     108      119   
29            dovenull        x     109      120   
30            postgrey        x     110      121   
31        debian-spamd        x     111      122   
32            memcache        x     113      124   
33              genadi        x    1002     1003   
34               shira        x    1003     1004   
35               atara        x    1004     1005   
36              shikma        x    1005     1006   
37               amotz        x    1006     1007   
38               mysql        x     114      125   
39              clamav        x     115      126   
40              amavis        x     116      127   
41            opendkim        x     117      128   
42        gitlab-redis        x     999     1009   
43         gitlab-psql        x     998     1010   
44                 git        x    1007     1008   
45           opendmarc        x     118      129   
46  dkim-milter-python        x     119      130   
47              deploy        x    1008     1011   
48               redis        x     112      123   

                                  name                      homedir  \
0                                 root                        /root   
1                               daemon                    /usr/sbin   
2                                  bin                         /bin   
3                                  sys                         /dev   
4                                 sync                         /bin   
5                                games                   /usr/games   
6                                  man               /var/cache/man   
7                                   lp               /var/spool/lpd   
8                                 mail                    /var/mail   
9                                 news              /var/spool/news   
10                                uucp              /var/spool/uucp   
11                               proxy                         /bin   
12                            www-data                     /var/www   
13                              backup                 /var/backups   
14                Mailing List Manager                    /var/list   
15                                ircd                /var/run/ircd   
16  Gnats Bug-Reporting System (admin)               /var/lib/gnats   
17                              nobody                 /nonexistent   
18                                 NaN                 /home/sys

In [88]:
ex18_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49 entries, 0 to 48
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   username  49 non-null     object
 1   password  49 non-null     object
 2   userid    49 non-null     int64 
 3   groupid   49 non-null     int64 
 4   name      35 non-null     object
 5   homedir   49 non-null     object
 6   shell     49 non-null     object
dtypes: int64(2), object(5)
memory usage: 2.8+ KB


# Exercise 19: Bitcoin values  

The closing price for the most recent trading day

The lowest historical price and the date of that price

The highest historical price and the date of that price

In [89]:
ex19_df = pd.read_csv('https://api.blockchain.info/charts/market-price?format=csv', 
                      header=None, 
                      names=['dates', 'value'])

ex19_df

dates     value
0    2023-10-25 00:00:00  33902.65
1    2023-10-26 00:00:00  34493.19
2    2023-10-27 00:00:00  34157.79
3    2023-10-28 00:00:00  33911.08
4    2023-10-29 00:00:00  34090.68
..                   ...       ...
361  2024-10-20 00:00:00  68365.74
362  2024-10-21 00:00:00  69012.04
363  2024-10-22 00:00:00  67363.26
364  2024-10-23 00:00:00  67368.20
365  2024-10-24 00:00:00  66655.68

[366 rows x 2 columns]

In [90]:
ex19_df.max() # this code will generate the latest date from column `dates`

dates    2024-10-24 00:00:00
value               73094.37
dtype: object

In [91]:
ex19_df.min() # this code will generate the oldest date from column `dates`

dates    2023-10-25 00:00:00
value               33902.65
dtype: object

In [92]:
ex19_df.loc[ex19_df['value'] == ex19_df['value'].min(), ['dates','value']]

dates     value
0  2023-10-25 00:00:00  33902.65

In [93]:
ex19_df.loc[ex19_df['value'] == ex19_df['value'].max(), ['dates','value']]

dates     value
141  2024-03-14 00:00:00  73094.37

In [94]:
ex19_df.head(1)

dates     value
0  2023-10-25 00:00:00  33902.65

In [95]:
ex19_df.tail(1)

dates     value
365  2024-10-24 00:00:00  66655.68

In [96]:
ex19_df.set_index('dates').idxmin()

value    2023-10-25 00:00:00
dtype: object

In [97]:
ex19_df.set_index('dates').idxmax()

value    2024-03-14 00:00:00
dtype: object

In [98]:
ex19_df.set_index('dates').agg(['idxmin', 'idxmax'])

value
idxmin  2023-10-25 00:00:00
idxmax  2024-03-14 00:00:00

### Beyong the exercise 
Pandas is full of amazing functionality that lets us retrieve data from the internet in various formats. Here are a few additional exercises for you to try to see how this works and how you can integrate it into your workflow:

In this exercise, you downloaded the information into a data frame and then performed calculations on it. Without assigning the downloaded data to an interim variable, can you return the current value? Your solution should consist of a single line of code that includes the download, selection, and calculation.

The pd.read_html function, like pd.read_csv, takes a file-like object or a URL. It assumes that it will encounter HTML-formatted text containing at least one table. It turns each table into a data frame and then returns a list of those data frames. With this in mind, retrieve one year of historical S&P 500 data from Yahoo Finance (https://finance.yahoo.com/quote/%5EGSPC/history?p=%5EGSPC), looking only at the Date, Close, and Volume columns. Show the date and volume of the days with the highest and lowest Close values. Note that Yahoo seems to look at the User-Agent header in the HTTP request, which cannot be set in read_html. So you’ll need to use requests to retrieve the data, setting User-Agent to a string equal to 'Mozilla 5.0'. Turn the content of the result into a StringIO, and then feed that to read_html and retrieve the data.

Create a two-row data frame with the highest and lowest closing prices for the S&P 500. Use the to_csv function to write this data to a new CSV file.

# Exercise 20: Big Cities 


### Reading a JSON file 

In [99]:
filepath20 = r"C:\Users\dqthi\Downloads\pandas-workout-data\data\cities.json"

In [100]:
ex20_df = pd.read_json(filepath20)

In [101]:
ex20_df

city growth_from_2000_to_2013   latitude   longitude  population  \
0        New York                     4.8%  40.712784  -74.005941     8405837   
1     Los Angeles                     4.8%  34.052234 -118.243685     3884307   
2         Chicago                    -6.1%  41.878114  -87.629798     2718782   
3         Houston                    11.0%  29.760427  -95.369803     2195914   
4    Philadelphia                     2.6%  39.952584  -75.165222     1553165   
..            ...                      ...        ...         ...         ...   
995       Weslaco                    28.8%  26.159519  -97.990837       37093   
996        Keizer                    14.4%  44.990119 -123.026208       37064   
997  Spanish Fork                    78.1%  40.114955 -111.654923       36956   
998        Beloit                     2.9%  42.508348  -89.031776       36888   
999   Panama City                     0.1%  30.158813  -85.660206       36877   

     rank         state  
0       1      New York  
1       2    California  
2       3      Illinois  
3       4         Texas  
4       5  Pennsylvania  
..    ...           ...  
995   996         Texas  
996   997        Oregon  
997   998          Utah  
998   999     Wisconsin  
999  1000       Florida  

[1000 rows x 7 columns]

In [102]:
ex20_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 7 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   city                      1000 non-null   object 
 1   growth_from_2000_to_2013  1000 non-null   object 
 2   latitude                  1000 non-null   float64
 3   longitude                 1000 non-null   float64
 4   population                1000 non-null   int64  
 5   rank                      1000 non-null   int64  
 6   state                     1000 non-null   object 
dtypes: float64(2), int64(2), object(3)
memory usage: 54.8+ KB


In [103]:
ex20_df['population'].describe()[['mean', '50%']]

mean    131132.443
50%      68207.000
Name: population, dtype: float64

In [104]:
ex20_df.loc[50:, 'population'].describe()[['mean', '50%']]

mean    87027.387368
50%     65796.000000
Name: population, dtype: float64

In [105]:
ex20_df.loc[ex20_df['latitude'] == ex20_df['latitude'].max(), ['city', 'state', 'rank' ]]

city   state  rank
62  Anchorage  Alaska    63

In [106]:
ex20_df['state'].value_counts().head(1) 

state
California    212
Name: count, dtype: int64

In [107]:
ex20_df['state'].value_counts().tail(5)

state
Alaska                  1
Hawaii                  1
District of Columbia    1
Maine                   1
Vermont                 1
Name: count, dtype: int64

### Beyond the exercise
Convert the growth_from_2000_to_2013 column into a floating-point number. Then find the mean and median changes in city size between 2000 and 2013. If a city has no recorded growth, set it to 0.

How many cities had positive growth in this period, and how many had negative growth?

Find the city or cities with latitudes more than two standard deviations from the mean.

In [108]:
ex20_df['growth_in_float'] = ex20_df['growth_from_2000_to_2013'].str.strip('%')

In [109]:
ex20_df['growth_in_float'] = ex20_df['growth_in_float'].astype(float) 

ValueError: could not convert string to float: ''

In [110]:
print(ex20_df[ex20_df['growth_from_2000_to_2013'] == ''])

                  city growth_from_2000_to_2013   latitude   longitude  \
299      Jurupa Valley                           33.997197 -117.485480   
656      Weymouth Town                           42.218072  -70.941036   
659           Eastvale                           33.952463 -117.584802   
728         Brookhaven                           33.865103  -84.336592   
783           Dunwoody                           33.946213  -84.334647   
922  Peachtree Corners                           33.969893  -84.221455   

     population  rank          state growth_in_float  
299       98030   300     California                  
656       55419   657  Massachusetts                  
659       55191   660     California                  
728       50603   729        Georgia                  
783       47591   784        Georgia                  
922       40059   923        Georgia                  


In [112]:
ex20_df.isna().sum()

city                        0
growth_from_2000_to_2013    0
latitude                    0
longitude                   0
population                  0
rank                        0
state                       0
growth_in_float             0
dtype: int64

In [117]:
to_drop_row = ex20_df[ex20_df['growth_from_2000_to_2013'] =='']

In [119]:
to_drop_row

city growth_from_2000_to_2013   latitude   longitude  \
299      Jurupa Valley                           33.997197 -117.485480   
656      Weymouth Town                           42.218072  -70.941036   
659           Eastvale                           33.952463 -117.584802   
728         Brookhaven                           33.865103  -84.336592   
783           Dunwoody                           33.946213  -84.334647   
922  Peachtree Corners                           33.969893  -84.221455   

     population  rank          state growth_in_float  
299       98030   300     California                  
656       55419   657  Massachusetts                  
659       55191   660     California                  
728       50603   729        Georgia                  
783       47591   784        Georgia                  
922       40059   923        Georgia

In [132]:
ex20b_df = ex20_df.drop(index=[299, 656, 659, 728, 783, 922])

In [133]:
ex20b_df = ex20b_df['growth_in_float'].astype(np.float16)

In [134]:
ex20b_df

0       4.800781
1       4.800781
2      -6.101562
3      11.000000
4       2.599609
         ...    
995    28.796875
996    14.398438
997    78.125000
998     2.900391
999     0.099976
Name: growth_in_float, Length: 994, dtype: float16

In [140]:
ex20_df[ex20_df['growth_from_2000_to_2013'] == ''] = 0

In [146]:
ex20_df['growth_in_float'] = ex20_df['growth_in_float'].astype(float)

In [147]:
ex20_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   city                      1000 non-null   object 
 1   growth_from_2000_to_2013  1000 non-null   object 
 2   latitude                  1000 non-null   float64
 3   longitude                 1000 non-null   float64
 4   population                1000 non-null   int64  
 5   rank                      1000 non-null   int64  
 6   state                     1000 non-null   object 
 7   growth_in_float           1000 non-null   float64
dtypes: float64(3), int64(2), object(3)
memory usage: 62.6+ KB


In [148]:
ex20_df['growth_in_float'].describe()

count    1000.000000
mean       22.936000
std        86.662216
min       -27.100000
25%         2.800000
50%         9.650000
75%        24.525000
max      2503.400000
Name: growth_in_float, dtype: float64

In [154]:
ex20_df.loc[ex20_df['growth_in_float'] > 0, 'city'].count()

847

In [153]:
ex20_df.loc[ex20_df['growth_in_float'] < 0, 'city'].count()

142